modules/cnn_3dobj/include/opencv2/cnn_3dobj.hpp

```diff
@@ -0,0 +1,252 @@
+indirect, incidental, special, exemplary, or consequential damages
+(including, but not limited to, procurement of substitute goods or services;
+loss of use, data, or profits; or business interruption) however caused
+and on any theory of liability, whether in contract, strict liability,
+or tort (including negligence or otherwise) arising in any way out of
+the use of this software, even if advised of the possibility of such damage.
+*/
+
+#ifndef __OPENCV_CNN_3DOBJ_HPP__
+#define __OPENCV_CNN_3DOBJ_HPP__
+#ifdef __cplusplus
+
+#include <string>
+#include <fstream>
+#include <vector>
+#include <stdio.h>
+#include <math.h>
+#include <iostream>
+#include <set>
+#include <string.h>
+#include <stdlib.h>
+#include <tr1/memory>
+#include <dirent.h>
+#define CPU_ONLY
+
+#ifdef HAVE_CAFFE
+#include <caffe/blob.hpp>
+#include <caffe/common.hpp>
+#include <caffe/net.hpp>
+#include <caffe/proto/caffe.pb.h>
+#include <caffe/util/io.hpp>
+#include <caffe/vision_layers.hpp>
+#endif
+
+#include "opencv2/viz/vizcore.hpp"
+#include "opencv2/highgui.hpp"
+#include "opencv2/highgui/highgui_c.h"
+#include "opencv2/imgproc.hpp"
+using std::string;
+using caffe::Blob;
+using caffe::Caffe;
+using caffe::Datum;
+using caffe::Net;
+/** @defgroup cnn_3dobj 3D object recognition and pose estimation API
+
+As CNN based learning algorithm shows better performance on the classification issues,
+the rich labeled data could be more useful in the training stage. 3D object classification and pose estimation
+is a jointed mission aimming at seperate different posed apart in the descriptor form.
+
+In the training stage, we prepare 2D training images generated from our module with their
+class label and pose label. We fully exploit the information lies in their labels
+by using a triplet and pair-wise jointed loss function in CNN training.
+Both class and pose label are in consideration in the triplet loss. The loss
+will be smaller when features from the same class and same pose is more similar
+and features from different classes or different poses are much larger.
+This loss is also jointed with a pair wise componant to make sure the loss is never be zero
+and have a restriction on the model scale.
+
+About the training and feature extraction process, it is a rough implemetation by using OpenCV
+and Caffe from the idea of Paul Wohlhart. The principal purpose of this API is construting
+a well labeled database from .ply models for CNN training with triplet loss and extracting features
+with the constructed model for prediction or other purpose of pattern recognition, algorithms into two main Class:
+
+**icoSphere**: methods belonging to this class generates a well generated database for CNN
+training together with their class and pose from camera view labels.
+
+**extractor**: methods belonging to this class extract features from 2D images which is
+discriminant on catogery prediction and pose estimation.
```

Comment:
replace:
As CNN based learning algorithm shows better performance on the classification issues,
the rich labeled data could be more useful in the training stage. 3D object classification and pose estimation
is a jointed mission aiming at separate different posea apart in the descriptor form.
In the training stage, we prepare 2D training images generated from our module with their
class label and pose label. We fully exploit the information that lies in their labels
by using a triplet and pair-wise jointed loss function in CNN training.
Both class and pose label are in consideration in the triplet loss. The loss score 
will be smaller when features from the same class and same pose is more similar
and features from different classes or different poses will lead to a much larger loss score.
This loss is also jointed with a pair wise component to make sure the loss is never be zero
and have a restriction on the model scale.
About the training and feature extraction process, it is a rough implementation by using OpenCV
and Caffe from the idea of Paul Wohlhart. The principal purpose of this API is constructing
a well labeled database from .ply models for CNN training with triplet loss and extracting features
with the constructed model for prediction or other purpose of pattern recognition, algorithms into two main Class:
**icoSphere**: methods belonging to this class generates 2D images from a 3D model, together with their class and pose from camera view labels.
**extractor**: methods belonging to this class extract descriptors from 2D images which is
discriminant on category prediction and pose estimation.